### PR TITLE
Gateway ext. enhancements | Search by REL Sync strategy performance optimizations

### LINKED ISSUE #78: Add Support for HTTP POST Requests in RequestMutation
**Description:**

We have previously encountered 400 errors when sending search requests to the JPA Server through the gateway-extensions. This issue was caused by the request size being too large. We resolved this by switching from GET to POST requests, as detailed in this issue: [Switch to HTTP POST for FHIR Search Requests #73](https://github.com/onaio/fhir-gateway-extension/issues/73). This change handled all search requests where we used the fhirclient.

However, there are sections that use `RequestMutation` provided by the fhir-gateway dependency used by the gateway-extension. These sections were not updated, resulting in the same 400 errors as reported in this Sentry issue: [Sentry Issue #49827](https://sentry.k8s.onalabs.org/organizations/ona/issues/49827/?project=145&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=14d&stream_index=0).

To resolve this issue, two tasks are required:

1. **Support for POST Requests in the fhir-gateway Repository:**
   Currently, the fhir-gateway does not support mutating HTTP POST requests. An open issue exists for this: [Request Mutation for POST Requests #143](https://github.com/google/fhir-gateway/issues/143). This support needs to be implemented in the fhir-gateway before changes can be made in the gateway-extensions repository.

2. **Implement Support for POST Requests in the gateway-extensions Repository:**
   Once the fhir-gateway supports mutating HTTP POST requests, we need to update the gateway-extensions to handle these requests correctly.

**Acceptance Criteria:**

- [ ] fhir-gateway repository supports mutating HTTP POST requests.
- [ ] gateway-extensions repository is updated to use POST requests in all sections that use `RequestMutation`.
- [ ] No more 400 errors due to large request sizes in sections using `RequestMutation`.

**References:**
- [Switch to HTTP POST for FHIR Search Requests #73](https://github.com/onaio/fhir-gateway-extension/issues/73)
- [Sentry Issue #49827](https://sentry.k8s.onalabs.org/organizations/ona/issues/49827/?project=145&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=14d&stream_index=0)
- [Request Mutation for POST Requests #143](https://github.com/google/fhir-gateway/issues/143)

### PR DESCRIPTION
**Pull request by ndegwamartin**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Resolves #78 

**Engineer Checklist**

- [x] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [x] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [x] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [x] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [x] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 1]
```diff
@@ -49,14 +50,15 @@ public class SyncAccessDecision implements AccessDecision {
...
-    private Map<String, List<String>> syncStrategyIdsMap;
+    private final Map<String, List<String>> syncStrategyIdsMap;
     private final List<String> roles;
     private IgnoredResourcesConfig config;
     private final String keycloakUUID;
     private final Gson gson = new Gson();
     private FhirContext fhirR4Context;
     private final IParser fhirR4JsonParser;
     private IGenericClient fhirR4Client;
+    private static final int REL_LOCATION_CHUNKSIZE = 20;
```
Comment (dubdabasoduba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:61]: Is this configurable?

Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:61]: No it isn't and I don't think it should be at the moment since this is where the Large URI Exception stems from. We want it to be the least highest possible value.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java [hunk 2]
```diff
@@ -222,6 +251,101 @@ public String postProcess(RequestDetailsReader request, HttpResponse response)
         return resultContent;
     }
 
+    private IBaseResource processRelatedEntityLocationSyncStrategy(
+            RequestDetailsReader request, HttpResponse response) throws IOException {
+        String resultContent;
+        fhirR4Client.getFhirContext().getRestfulClientFactory().setConnectionRequestTimeout(300000);
+        fhirR4Client.getFhirContext().getRestfulClientFactory().setSocketTimeout(300000);
+
+        int subListSize = 100;
```
[the remark this answers is not in the document]
Comment (ndegwamartin) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java:260]: Cool let me move it to the inner Constants class at the bottom. Should be internal to the class and also should not be re-configured so as not to impact the proxy to FHIR server communication _(Based on tests we did with SRE this was the optimal pagination value for server resources deployed for ESM MG Preview)_

We could however make it configurable but with a caveat on how it could impact stability.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java (hunk 1 above)
Comment (dubdabasoduba) [reply plugins/src/main/java/org/smartregister/fhir/gateway/plugins/SyncAccessDecision.java]: Makes sense

**Review by dubdabasoduba — APPROVED** [review]
